# Management - Fragmenter Validation 1.0

In [1]:
import qcportal as ptl
import pandas as pd
import datetime
import time

from management import *

In [2]:
# connect with authentication, therefore write access
# don't use unless you plan to submit things
client = ptl.FractalClient.from_file()

In [3]:
client

FractalClient(server_name='The MolSSI QCArchive Server', address='https://api.qcarchive.molssi.org:443/', username='ddotson')

## Dataset tracking

### 2020.06.27 01:00 UTC

In [4]:
dataset = "OpenFF Fragmenter Validation 1.0"

In [5]:
ds = client.get_collection("TorsionDriveDataset", dataset)

In [7]:
ds.list_specifications()

Description
Name                                                                     
default                       Standard OpenFF torsiondrive specification.
B3LYP-d3bj/aug-cc-pvdz  OpenFF model B3LYP-d3bj/aug-cc-pvdz TorsionDri...
B3LYP-d3bj/aug-cc-pvtz  OpenFF model B3LYP-d3bj/aug-cc-pvtz TorsionDri...
B3LYP-d3bj/def2-tzvppd  OpenFF model B3LYP-d3bj/def2-tzvppd TorsionDri...
BLYP-d3bj/dzvp          OpenFF model BLYP-d3bj/dzvp TorsionDrive explo...
BLYP-d3bj/aug-cc-pvdz   OpenFF model BLYP-d3bj/aug-cc-pvdz TorsionDriv...
BLYP-d3bj/aug-cc-pvtz   OpenFF model BLYP-d3bj/aug-cc-pvtz TorsionDriv...
BLYP-d3bj/def2-tzvppd   OpenFF model BLYP-d3bj/def2-tzvppd TorsionDriv...
LRC-WPBEH/dzvp          OpenFF model LRC-WPBEH/dzvp TorsionDrive explo...
LRC-WPBEH/aug-cc-pvdz   OpenFF model LRC-WPBEH/aug-cc-pvdz TorsionDriv...
LRC-WPBEH/aug-cc-pvtz   OpenFF model LRC-WPBEH/aug-cc-pvtz TorsionDriv...
LRC-WPBEH/def2-tzvppd   OpenFF model LRC-WPBEH/def2-tzvppd TorsionDriv...
MN15-d3bJ/dzvp          OpenFF model MN15-d3bJ/dzvp TorsionDrive explo...
MN15-d3bJ/aug-cc-pvdz   OpenFF model MN15-d3bJ/aug-cc-pvdz TorsionDriv...
MN15-d3bJ/aug-cc-pvtz   OpenFF model MN15-d3bJ/aug-cc-pvtz TorsionDriv...
MN15-d3bJ/def2-tzvppd   OpenFF model MN15-d3bJ/def2-tzvppd TorsionDriv...
TPSSH-d3bJ/dzvp         OpenFF model TPSSH-d3bJ/dzvp TorsionDrive expl...
TPSSH-d3bJ/aug-cc-pvdz  OpenFF model TPSSH-d3bJ/aug-cc-pvdz TorsionDri...
TPSSH-d3bJ/aug-cc-pvtz  OpenFF model TPSSH-d3bJ/aug-cc-pvtz TorsionDri...
TPSSH-d3bJ/def2-tzvppd  OpenFF model TPSSH-d3bJ/def2-tzvppd TorsionDri...

In [9]:
res = get_unfinished_torsiondrive_optimizations(ds, 'default', client)
res

{'20548975': [OptimizationRecord(id='20549211', status='COMPLETE'),
  OptimizationRecord(id='20550249', status='COMPLETE'),
  OptimizationRecord(id='20550250', status='COMPLETE'),
  OptimizationRecord(id='20556649', status='COMPLETE'),
  OptimizationRecord(id='20556650', status='COMPLETE'),
  OptimizationRecord(id='20556651', status='COMPLETE'),
  OptimizationRecord(id='20556652', status='COMPLETE'),
  OptimizationRecord(id='20657510', status='COMPLETE'),
  OptimizationRecord(id='20657511', status='COMPLETE'),
  OptimizationRecord(id='20657512', status='COMPLETE'),
  OptimizationRecord(id='20657513', status='COMPLETE'),
  OptimizationRecord(id='20658692', status='COMPLETE'),
  OptimizationRecord(id='20658693', status='COMPLETE'),
  OptimizationRecord(id='20658694', status='COMPLETE'),
  OptimizationRecord(id='20658695', status='COMPLETE'),
  OptimizationRecord(id='20659679', status='COMPLETE'),
  OptimizationRecord(id='20659680', status='COMPLETE'),
  OptimizationRecord(id='20659681', 

In [11]:
resm = merge(res)

In [28]:
incomplete = (opt.dict() for opt in resm if opt.status == 'INCOMPLETE')

In [29]:
next(incomplete)

{'id': '20867730',
 'hash_index': '13a44bf160e1ef987c807575b3ec18e13cdfe0e5',
 'procedure': 'optimization',
 'program': 'geometric',
 'version': 1,
 'protocols': {},
 'extras': {},
 'stdout': '24143049',
 'stderr': None,
 'error': None,
 'task_id': None,
 'manager_name': None,
 'status': <RecordStatusEnum.incomplete: 'INCOMPLETE'>,
 'modified_on': datetime.datetime(2020, 5, 22, 13, 6, 15, 97769),
 'created_on': datetime.datetime(2020, 5, 22, 13, 6, 15, 97771),
 'provenance': {'creator': 'geomeTRIC',
  'version': '0.9.7.2',
  'routine': 'geometric.run_json.geometric_run_json',
  'cpu': 'Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz',
  'wall_time': 368.8933312892914,
  'hostname': 'lu05',
  'qcengine_version': '0.14.0',
  'username': 'chodera'},
 'schema_version': 1,
 'initial_molecule': '14375471',
 'qc_spec': {'driver': <DriverEnum.gradient: 'gradient'>,
  'method': 'b3lyp-d3bj',
  'basis': 'dzvp',
  'keywords': '2',
  'program': 'psi4'},
 'keywords': {'coordsys': 'tric',
  'enforce': 0.1

All Lilac INCOMPLETEs. INCOMPLETEs are over a week old. We need to clear these out somehow with Ben. I anticipate this needing clearing on the server.

In [30]:
count_unique_optimization_error_messages(resm, full=True, pretty_print=True)

There are 4 instances of
-------------------------------------
geomeTRIC run_json error:
Traceback (most recent call last):
  File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/run_json.py", line 225, in geometric_run_json
    geometric.optimize.Optimize(coords, M, IC, engine, None, params)
  File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/optimize.py", line 1331, in Optimize
    return optimizer.optimizeGeometry()
  File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/optimize.py", line 1298, in optimizeGeometry
    self.calcEnergyForce()
  File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/optimize.py", line 1002, in calcEnergyForce
    spcalc = self.engine.calc(self.X, self.dirname)
  File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/engine.py", line 873, in calc
    re

In [31]:
restart_optimizations(resm, client)

OptimizationRecord(id='19819414', status='ERROR')
OptimizationRecord(id='20549279', status='ERROR')
OptimizationRecord(id='20820900', status='ERROR')
OptimizationRecord(id='20820902', status='ERROR')


In [32]:
[tdr for tdr in ds.df.default if tdr.status == 'ERROR']

[TorsionDriveRecord(id='19712170', status='ERROR'),
 TorsionDriveRecord(id='20548997', status='ERROR'),
 TorsionDriveRecord(id='20548998', status='ERROR')]

In [33]:
restart_torsiondrives(ds.df.default.tolist(), client)

TorsionDriveRecord(id='19712170', status='ERROR')
TorsionDriveRecord(id='20548997', status='ERROR')
TorsionDriveRecord(id='20548998', status='ERROR')


### 2020.06.18 14:10 UTC

In [4]:
dataset = "OpenFF Fragmenter Validation 1.0"

In [5]:
ds = client.get_collection("TorsionDriveDataset", dataset)

In [6]:
ds.list_specifications()

Description
Name                                                                     
default                       Standard OpenFF torsiondrive specification.
B3LYP-d3bj/aug-cc-pvdz  OpenFF model B3LYP-d3bj/aug-cc-pvdz TorsionDri...
B3LYP-d3bj/aug-cc-pvtz  OpenFF model B3LYP-d3bj/aug-cc-pvtz TorsionDri...
B3LYP-d3bj/def2-tzvppd  OpenFF model B3LYP-d3bj/def2-tzvppd TorsionDri...
BLYP-d3bj/dzvp          OpenFF model BLYP-d3bj/dzvp TorsionDrive explo...
BLYP-d3bj/aug-cc-pvdz   OpenFF model BLYP-d3bj/aug-cc-pvdz TorsionDriv...
BLYP-d3bj/aug-cc-pvtz   OpenFF model BLYP-d3bj/aug-cc-pvtz TorsionDriv...
BLYP-d3bj/def2-tzvppd   OpenFF model BLYP-d3bj/def2-tzvppd TorsionDriv...
LRC-WPBEH/dzvp          OpenFF model LRC-WPBEH/dzvp TorsionDrive explo...
LRC-WPBEH/aug-cc-pvdz   OpenFF model LRC-WPBEH/aug-cc-pvdz TorsionDriv...
LRC-WPBEH/aug-cc-pvtz   OpenFF model LRC-WPBEH/aug-cc-pvtz TorsionDriv...
LRC-WPBEH/def2-tzvppd   OpenFF model LRC-WPBEH/def2-tzvppd TorsionDriv...
MN15-d3bJ/dzvp          OpenFF model MN15-d3bJ/dzvp TorsionDrive explo...
MN15-d3bJ/aug-cc-pvdz   OpenFF model MN15-d3bJ/aug-cc-pvdz TorsionDriv...
MN15-d3bJ/aug-cc-pvtz   OpenFF model MN15-d3bJ/aug-cc-pvtz TorsionDriv...
MN15-d3bJ/def2-tzvppd   OpenFF model MN15-d3bJ/def2-tzvppd TorsionDriv...
TPSSH-d3bJ/dzvp         OpenFF model TPSSH-d3bJ/dzvp TorsionDrive expl...
TPSSH-d3bJ/aug-cc-pvdz  OpenFF model TPSSH-d3bJ/aug-cc-pvdz TorsionDri...
TPSSH-d3bJ/aug-cc-pvtz  OpenFF model TPSSH-d3bJ/aug-cc-pvtz TorsionDri...
TPSSH-d3bJ/def2-tzvppd  OpenFF model TPSSH-d3bJ/def2-tzvppd TorsionDri...

In [7]:
res = get_unfinished_torsiondrive_optimizations(ds, 'default', client, merged=False)
res

{'20548975': [OptimizationRecord(id='20549211', status='COMPLETE'),
  OptimizationRecord(id='20550249', status='COMPLETE'),
  OptimizationRecord(id='20550250', status='COMPLETE'),
  OptimizationRecord(id='20556649', status='COMPLETE'),
  OptimizationRecord(id='20556650', status='COMPLETE'),
  OptimizationRecord(id='20556651', status='COMPLETE'),
  OptimizationRecord(id='20556652', status='COMPLETE'),
  OptimizationRecord(id='20657510', status='COMPLETE'),
  OptimizationRecord(id='20657511', status='COMPLETE'),
  OptimizationRecord(id='20657512', status='COMPLETE'),
  OptimizationRecord(id='20657513', status='COMPLETE'),
  OptimizationRecord(id='20658692', status='COMPLETE'),
  OptimizationRecord(id='20658693', status='COMPLETE'),
  OptimizationRecord(id='20658694', status='COMPLETE'),
  OptimizationRecord(id='20658695', status='COMPLETE'),
  OptimizationRecord(id='20659679', status='COMPLETE'),
  OptimizationRecord(id='20659680', status='COMPLETE'),
  OptimizationRecord(id='20659681', 

In [8]:
resm = merge(res)

In [42]:
incomplete = (opt.dict() for opt in resm if opt.status == 'INCOMPLETE')

In [49]:
next(incomplete)

{'id': '20321335',
 'hash_index': '4c50828aa750287753274d21d0f14248989141a2',
 'procedure': 'optimization',
 'program': 'geometric',
 'version': 1,
 'protocols': {},
 'extras': {},
 'stdout': '23627523',
 'stderr': None,
 'error': None,
 'task_id': None,
 'manager_name': None,
 'status': <RecordStatusEnum.incomplete: 'INCOMPLETE'>,
 'modified_on': datetime.datetime(2020, 5, 14, 16, 52, 38, 501432),
 'created_on': datetime.datetime(2020, 5, 14, 2, 22, 4, 348635),
 'provenance': {'creator': 'geomeTRIC',
  'version': '0.9.7.2',
  'routine': 'geometric.run_json.geometric_run_json',
  'username': 'chodera',
  'cpu': 'Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz',
  'qcengine_version': '0.14.0',
  'hostname': 'lt21',
  'wall_time': 1576.063833475113},
 'schema_version': 1,
 'initial_molecule': '13855947',
 'qc_spec': {'driver': <DriverEnum.gradient: 'gradient'>,
  'method': 'b3lyp-d3bj',
  'basis': 'dzvp',
  'keywords': '2',
  'program': 'psi4'},
 'keywords': {'coordsys': 'tric',
  'enforce': 0

Not all Lilac INCOMPLETEs. Other INCOMPLETEs are over a week old. We need to clear these out somehow with Ben. I anticipate this needing clearing on the server.

In [29]:
count_unique_optimization_error_messages(resm, full=True, pretty_print=True)

In [30]:
restart_optimizations(resm, client)

In [31]:
[tdr for tdr in ds.df.default if tdr.status == 'ERROR']

[]

In [32]:
restart_torsiondrives(ds.df.default.tolist(), client)

No current errors!

### 2020.06.16 15:10 UTC

In [4]:
dataset = "OpenFF Fragmenter Validation 1.0"

In [5]:
ds = client.get_collection("TorsionDriveDataset", dataset)

In [6]:
ds.list_specifications()

Description
Name                                                                     
default                       Standard OpenFF torsiondrive specification.
B3LYP-d3bj/aug-cc-pvdz  OpenFF model B3LYP-d3bj/aug-cc-pvdz TorsionDri...
B3LYP-d3bj/aug-cc-pvtz  OpenFF model B3LYP-d3bj/aug-cc-pvtz TorsionDri...
B3LYP-d3bj/def2-tzvppd  OpenFF model B3LYP-d3bj/def2-tzvppd TorsionDri...
BLYP-d3bj/dzvp          OpenFF model BLYP-d3bj/dzvp TorsionDrive explo...
BLYP-d3bj/aug-cc-pvdz   OpenFF model BLYP-d3bj/aug-cc-pvdz TorsionDriv...
BLYP-d3bj/aug-cc-pvtz   OpenFF model BLYP-d3bj/aug-cc-pvtz TorsionDriv...
BLYP-d3bj/def2-tzvppd   OpenFF model BLYP-d3bj/def2-tzvppd TorsionDriv...
LRC-WPBEH/dzvp          OpenFF model LRC-WPBEH/dzvp TorsionDrive explo...
LRC-WPBEH/aug-cc-pvdz   OpenFF model LRC-WPBEH/aug-cc-pvdz TorsionDriv...
LRC-WPBEH/aug-cc-pvtz   OpenFF model LRC-WPBEH/aug-cc-pvtz TorsionDriv...
LRC-WPBEH/def2-tzvppd   OpenFF model LRC-WPBEH/def2-tzvppd TorsionDriv...
MN15-d3bJ/dzvp          OpenFF model MN15-d3bJ/dzvp TorsionDrive explo...
MN15-d3bJ/aug-cc-pvdz   OpenFF model MN15-d3bJ/aug-cc-pvdz TorsionDriv...
MN15-d3bJ/aug-cc-pvtz   OpenFF model MN15-d3bJ/aug-cc-pvtz TorsionDriv...
MN15-d3bJ/def2-tzvppd   OpenFF model MN15-d3bJ/def2-tzvppd TorsionDriv...
TPSSH-d3bJ/dzvp         OpenFF model TPSSH-d3bJ/dzvp TorsionDrive expl...
TPSSH-d3bJ/aug-cc-pvdz  OpenFF model TPSSH-d3bJ/aug-cc-pvdz TorsionDri...
TPSSH-d3bJ/aug-cc-pvtz  OpenFF model TPSSH-d3bJ/aug-cc-pvtz TorsionDri...
TPSSH-d3bJ/def2-tzvppd  OpenFF model TPSSH-d3bJ/def2-tzvppd TorsionDri...

In [7]:
res = get_unfinished_torsiondrive_optimizations(ds, 'default', client, merged=False)
res

{'20548975': [OptimizationRecord(id='20549211', status='COMPLETE'),
  OptimizationRecord(id='20550249', status='COMPLETE'),
  OptimizationRecord(id='20550250', status='COMPLETE'),
  OptimizationRecord(id='20556649', status='COMPLETE'),
  OptimizationRecord(id='20556650', status='COMPLETE'),
  OptimizationRecord(id='20556651', status='COMPLETE'),
  OptimizationRecord(id='20556652', status='COMPLETE'),
  OptimizationRecord(id='20657510', status='COMPLETE'),
  OptimizationRecord(id='20657511', status='COMPLETE'),
  OptimizationRecord(id='20657512', status='COMPLETE'),
  OptimizationRecord(id='20657513', status='COMPLETE'),
  OptimizationRecord(id='20658692', status='COMPLETE'),
  OptimizationRecord(id='20658693', status='COMPLETE'),
  OptimizationRecord(id='20658694', status='COMPLETE'),
  OptimizationRecord(id='20658695', status='COMPLETE'),
  OptimizationRecord(id='20659679', status='COMPLETE'),
  OptimizationRecord(id='20659680', status='COMPLETE'),
  OptimizationRecord(id='20659681', 

In [8]:
resm = merge(res)

In [19]:
incomplete = (opt.dict() for opt in resm if opt.status == 'INCOMPLETE')

In [24]:
next(incomplete)

{'id': '20820900',
 'hash_index': 'b783f19b7e1ae18aeef18139394de77b4388411d',
 'procedure': 'optimization',
 'program': 'geometric',
 'version': 1,
 'protocols': {},
 'extras': {},
 'stdout': None,
 'stderr': None,
 'error': '24092960',
 'task_id': None,
 'manager_name': 'OpenFF_Mobley_HPC3-login-i16-1f98e0f3-0122-4a49-9347-577ff3048786',
 'status': <RecordStatusEnum.incomplete: 'INCOMPLETE'>,
 'modified_on': datetime.datetime(2020, 6, 12, 15, 53, 37, 553791),
 'created_on': datetime.datetime(2020, 5, 21, 9, 44, 12, 444427),
 'provenance': None,
 'schema_version': 1,
 'initial_molecule': '14342036',
 'qc_spec': {'driver': <DriverEnum.gradient: 'gradient'>,
  'method': 'b3lyp-d3bj',
  'basis': 'dzvp',
  'keywords': '2',
  'program': 'psi4'},
 'keywords': {'coordsys': 'tric',
  'enforce': 0.1,
  'reset': True,
  'qccnv': True,
  'epsilon': 0,
  'constraints': {'set': [{'type': 'dihedral',
     'indices': [8, 12, 9, 11],
     'value': 45}]},
  'program': 'psi4'},
 'energies': None,
 'fina

Not all Lilac INCOMPLETEs. Some of these may be legit PRP/UCI INCOMPLETEs.

In [25]:
count_unique_optimization_error_messages(resm, full=True, pretty_print=True)

In [26]:
restart_optimizations(resm, client)

In [29]:
[tdr for tdr in ds.df.default if tdr.status == 'ERROR']

[]

In [30]:
restart_torsiondrives(ds.df.default.tolist(), client)

No current errors!

### 2020.06.13 22:45 UTC

In [6]:
dataset = "OpenFF Fragmenter Validation 1.0"

In [7]:
ds = client.get_collection("TorsionDriveDataset", dataset)

In [9]:
ds.list_specifications()

Description
Name                                                                     
default                       Standard OpenFF torsiondrive specification.
B3LYP-d3bj/aug-cc-pvdz  OpenFF model B3LYP-d3bj/aug-cc-pvdz TorsionDri...
B3LYP-d3bj/aug-cc-pvtz  OpenFF model B3LYP-d3bj/aug-cc-pvtz TorsionDri...
B3LYP-d3bj/def2-tzvppd  OpenFF model B3LYP-d3bj/def2-tzvppd TorsionDri...
BLYP-d3bj/dzvp          OpenFF model BLYP-d3bj/dzvp TorsionDrive explo...
BLYP-d3bj/aug-cc-pvdz   OpenFF model BLYP-d3bj/aug-cc-pvdz TorsionDriv...
BLYP-d3bj/aug-cc-pvtz   OpenFF model BLYP-d3bj/aug-cc-pvtz TorsionDriv...
BLYP-d3bj/def2-tzvppd   OpenFF model BLYP-d3bj/def2-tzvppd TorsionDriv...
LRC-WPBEH/dzvp          OpenFF model LRC-WPBEH/dzvp TorsionDrive explo...
LRC-WPBEH/aug-cc-pvdz   OpenFF model LRC-WPBEH/aug-cc-pvdz TorsionDriv...
LRC-WPBEH/aug-cc-pvtz   OpenFF model LRC-WPBEH/aug-cc-pvtz TorsionDriv...
LRC-WPBEH/def2-tzvppd   OpenFF model LRC-WPBEH/def2-tzvppd TorsionDriv...
MN15-d3bJ/dzvp          OpenFF model MN15-d3bJ/dzvp TorsionDrive explo...
MN15-d3bJ/aug-cc-pvdz   OpenFF model MN15-d3bJ/aug-cc-pvdz TorsionDriv...
MN15-d3bJ/aug-cc-pvtz   OpenFF model MN15-d3bJ/aug-cc-pvtz TorsionDriv...
MN15-d3bJ/def2-tzvppd   OpenFF model MN15-d3bJ/def2-tzvppd TorsionDriv...
TPSSH-d3bJ/dzvp         OpenFF model TPSSH-d3bJ/dzvp TorsionDrive expl...
TPSSH-d3bJ/aug-cc-pvdz  OpenFF model TPSSH-d3bJ/aug-cc-pvdz TorsionDri...
TPSSH-d3bJ/aug-cc-pvtz  OpenFF model TPSSH-d3bJ/aug-cc-pvtz TorsionDri...
TPSSH-d3bJ/def2-tzvppd  OpenFF model TPSSH-d3bJ/def2-tzvppd TorsionDri...

In [10]:
res = get_unfinished_torsiondrive_optimizations(ds, 'default', client, merged=False)
res

{'20548975': [OptimizationRecord(id='20549211', status='COMPLETE'),
  OptimizationRecord(id='20550249', status='COMPLETE'),
  OptimizationRecord(id='20550250', status='COMPLETE'),
  OptimizationRecord(id='20556649', status='COMPLETE'),
  OptimizationRecord(id='20556650', status='COMPLETE'),
  OptimizationRecord(id='20556651', status='COMPLETE'),
  OptimizationRecord(id='20556652', status='COMPLETE'),
  OptimizationRecord(id='20657510', status='COMPLETE'),
  OptimizationRecord(id='20657511', status='COMPLETE'),
  OptimizationRecord(id='20657512', status='COMPLETE'),
  OptimizationRecord(id='20657513', status='COMPLETE'),
  OptimizationRecord(id='20658692', status='COMPLETE'),
  OptimizationRecord(id='20658693', status='COMPLETE'),
  OptimizationRecord(id='20658694', status='COMPLETE'),
  OptimizationRecord(id='20658695', status='COMPLETE'),
  OptimizationRecord(id='20659679', status='COMPLETE'),
  OptimizationRecord(id='20659680', status='COMPLETE'),
  OptimizationRecord(id='20659681', 

In [11]:
resm = merge(res)

In [34]:
incomplete = (opt.dict() for opt in resm if opt.status == 'INCOMPLETE')

In [35]:
next(incomplete)

{'id': '20867730',
 'hash_index': '13a44bf160e1ef987c807575b3ec18e13cdfe0e5',
 'procedure': 'optimization',
 'program': 'geometric',
 'version': 1,
 'protocols': {},
 'extras': {},
 'stdout': '24143049',
 'stderr': None,
 'error': None,
 'task_id': None,
 'manager_name': None,
 'status': <RecordStatusEnum.incomplete: 'INCOMPLETE'>,
 'modified_on': datetime.datetime(2020, 5, 22, 13, 6, 15, 97769),
 'created_on': datetime.datetime(2020, 5, 22, 13, 6, 15, 97771),
 'provenance': {'creator': 'geomeTRIC',
  'version': '0.9.7.2',
  'routine': 'geometric.run_json.geometric_run_json',
  'cpu': 'Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz',
  'wall_time': 368.8933312892914,
  'username': 'chodera',
  'qcengine_version': '0.14.0',
  'hostname': 'lu05'},
 'schema_version': 1,
 'initial_molecule': '14375471',
 'qc_spec': {'driver': <DriverEnum.gradient: 'gradient'>,
  'method': 'b3lyp-d3bj',
  'basis': 'dzvp',
  'keywords': '2',
  'program': 'psi4'},
 'keywords': {'coordsys': 'tric',
  'enforce': 0.1

Not all Lilac INCOMPLETEs. Some of these may be legit PRP/UCI INCOMPLETEs.

In [30]:
count_unique_optimization_error_messages(resm, full=True, pretty_print=True)

In [31]:
restart_optimizations(resm, client)

In [32]:
[tdr for tdr in ds.df.default if tdr.status == 'ERROR']

[]

In [33]:
restart_torsiondrives(ds.df.default.tolist(), client)

No current errors!

### 2020.06.12 15:30 UTC

In [7]:
dataset = "OpenFF Fragmenter Validation 1.0"

In [8]:
ds = client.get_collection("TorsionDriveDataset", dataset)

In [9]:
ds.list_specifications()

Description
Name                                                                     
default                       Standard OpenFF torsiondrive specification.
B3LYP-d3bj/aug-cc-pvdz  OpenFF model B3LYP-d3bj/aug-cc-pvdz TorsionDri...
B3LYP-d3bj/aug-cc-pvtz  OpenFF model B3LYP-d3bj/aug-cc-pvtz TorsionDri...
B3LYP-d3bj/def2-tzvppd  OpenFF model B3LYP-d3bj/def2-tzvppd TorsionDri...
BLYP-d3bj/dzvp          OpenFF model BLYP-d3bj/dzvp TorsionDrive explo...
BLYP-d3bj/aug-cc-pvdz   OpenFF model BLYP-d3bj/aug-cc-pvdz TorsionDriv...
BLYP-d3bj/aug-cc-pvtz   OpenFF model BLYP-d3bj/aug-cc-pvtz TorsionDriv...
BLYP-d3bj/def2-tzvppd   OpenFF model BLYP-d3bj/def2-tzvppd TorsionDriv...
LRC-WPBEH/dzvp          OpenFF model LRC-WPBEH/dzvp TorsionDrive explo...
LRC-WPBEH/aug-cc-pvdz   OpenFF model LRC-WPBEH/aug-cc-pvdz TorsionDriv...
LRC-WPBEH/aug-cc-pvtz   OpenFF model LRC-WPBEH/aug-cc-pvtz TorsionDriv...
LRC-WPBEH/def2-tzvppd   OpenFF model LRC-WPBEH/def2-tzvppd TorsionDriv...
MN15-d3bJ/dzvp          OpenFF model MN15-d3bJ/dzvp TorsionDrive explo...
MN15-d3bJ/aug-cc-pvdz   OpenFF model MN15-d3bJ/aug-cc-pvdz TorsionDriv...
MN15-d3bJ/aug-cc-pvtz   OpenFF model MN15-d3bJ/aug-cc-pvtz TorsionDriv...
MN15-d3bJ/def2-tzvppd   OpenFF model MN15-d3bJ/def2-tzvppd TorsionDriv...
TPSSH-d3bJ/dzvp         OpenFF model TPSSH-d3bJ/dzvp TorsionDrive expl...
TPSSH-d3bJ/aug-cc-pvdz  OpenFF model TPSSH-d3bJ/aug-cc-pvdz TorsionDri...
TPSSH-d3bJ/aug-cc-pvtz  OpenFF model TPSSH-d3bJ/aug-cc-pvtz TorsionDri...
TPSSH-d3bJ/def2-tzvppd  OpenFF model TPSSH-d3bJ/def2-tzvppd TorsionDri...

In [10]:
res = get_unfinished_torsiondrive_optimizations(ds, 'default', client, merged=False)
res

{'20548975': [OptimizationRecord(id='20549211', status='COMPLETE'),
  OptimizationRecord(id='20550249', status='COMPLETE'),
  OptimizationRecord(id='20550250', status='COMPLETE'),
  OptimizationRecord(id='20556649', status='COMPLETE'),
  OptimizationRecord(id='20556650', status='COMPLETE'),
  OptimizationRecord(id='20556651', status='COMPLETE'),
  OptimizationRecord(id='20556652', status='COMPLETE'),
  OptimizationRecord(id='20657510', status='COMPLETE'),
  OptimizationRecord(id='20657511', status='COMPLETE'),
  OptimizationRecord(id='20657512', status='COMPLETE'),
  OptimizationRecord(id='20657513', status='COMPLETE'),
  OptimizationRecord(id='20658692', status='COMPLETE'),
  OptimizationRecord(id='20658693', status='COMPLETE'),
  OptimizationRecord(id='20658694', status='COMPLETE'),
  OptimizationRecord(id='20658695', status='COMPLETE'),
  OptimizationRecord(id='20659679', status='COMPLETE'),
  OptimizationRecord(id='20659680', status='COMPLETE'),
  OptimizationRecord(id='20659681', 

In [12]:
resm = merge(res)

In [34]:
incomplete = (opt.dict() for opt in resm if opt.status == 'INCOMPLETE')

In [35]:
next(incomplete)

{'id': '20867730',
 'hash_index': '13a44bf160e1ef987c807575b3ec18e13cdfe0e5',
 'procedure': 'optimization',
 'program': 'geometric',
 'version': 1,
 'protocols': {},
 'extras': {},
 'stdout': '24143049',
 'stderr': None,
 'error': None,
 'task_id': None,
 'manager_name': None,
 'status': <RecordStatusEnum.incomplete: 'INCOMPLETE'>,
 'modified_on': datetime.datetime(2020, 5, 22, 13, 6, 15, 97769),
 'created_on': datetime.datetime(2020, 5, 22, 13, 6, 15, 97771),
 'provenance': {'creator': 'geomeTRIC',
  'version': '0.9.7.2',
  'routine': 'geometric.run_json.geometric_run_json',
  'cpu': 'Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz',
  'username': 'chodera',
  'qcengine_version': '0.14.0',
  'wall_time': 368.8933312892914,
  'hostname': 'lu05'},
 'schema_version': 1,
 'initial_molecule': '14375471',
 'qc_spec': {'driver': <DriverEnum.gradient: 'gradient'>,
  'method': 'b3lyp-d3bj',
  'basis': 'dzvp',
  'keywords': '2',
  'program': 'psi4'},
 'keywords': {'coordsys': 'tric',
  'enforce': 0.1

All Lilac INCOMPLETEs?

In [23]:
count_unique_optimization_error_messages(resm, full=True, pretty_print=True)

There are 1 instances of
-------------------------------------
geomeTRIC run_json error:
Traceback (most recent call last):
  File "/opt/conda/envs/qcfractal/lib/python3.6/site-packages/geometric/run_json.py", line 225, in geometric_run_json
    geometric.optimize.Optimize(coords, M, IC, engine, None, params)
  File "/opt/conda/envs/qcfractal/lib/python3.6/site-packages/geometric/optimize.py", line 1331, in Optimize
    return optimizer.optimizeGeometry()
  File "/opt/conda/envs/qcfractal/lib/python3.6/site-packages/geometric/optimize.py", line 1298, in optimizeGeometry
    self.calcEnergyForce()
  File "/opt/conda/envs/qcfractal/lib/python3.6/site-packages/geometric/optimize.py", line 1002, in calcEnergyForce
    spcalc = self.engine.calc(self.X, self.dirname)
  File "/opt/conda/envs/qcfractal/lib/python3.6/site-packages/geometric/engine.py", line 873, in calc
    return self.calc_new(coords, dirname)
  File "/opt/conda/envs/qcfractal/lib/python3.6/site-packages/geometric/engine.py", 

In [20]:
restart_optimizations(resm, client)

OptimizationRecord(id='19819414', status='ERROR')
OptimizationRecord(id='20549279', status='ERROR')
OptimizationRecord(id='20820900', status='ERROR')
OptimizationRecord(id='20820902', status='ERROR')


In [21]:
[tdr for tdr in ds.df.default if tdr.status == 'ERROR']

[TorsionDriveRecord(id='19712170', status='ERROR'),
 TorsionDriveRecord(id='20548997', status='ERROR'),
 TorsionDriveRecord(id='20548998', status='ERROR')]

In [22]:
restart_torsiondrives(ds.df.default.tolist(), client)

TorsionDriveRecord(id='19712170', status='ERROR')
TorsionDriveRecord(id='20548997', status='ERROR')
TorsionDriveRecord(id='20548998', status='ERROR')


--------
## Archive

In [5]:
ds = client.get_collection("TorsionDriveDataset", "OpenFF Fragmenter Validation 1.0")

In [6]:
ds.list_specifications()

Description
Name                                                                     
default                       Standard OpenFF torsiondrive specification.
B3LYP-d3bj/aug-cc-pvdz  OpenFF model B3LYP-d3bj/aug-cc-pvdz TorsionDri...
B3LYP-d3bj/aug-cc-pvtz  OpenFF model B3LYP-d3bj/aug-cc-pvtz TorsionDri...
B3LYP-d3bj/def2-tzvppd  OpenFF model B3LYP-d3bj/def2-tzvppd TorsionDri...
BLYP-d3bj/dzvp          OpenFF model BLYP-d3bj/dzvp TorsionDrive explo...
BLYP-d3bj/aug-cc-pvdz   OpenFF model BLYP-d3bj/aug-cc-pvdz TorsionDriv...
BLYP-d3bj/aug-cc-pvtz   OpenFF model BLYP-d3bj/aug-cc-pvtz TorsionDriv...
BLYP-d3bj/def2-tzvppd   OpenFF model BLYP-d3bj/def2-tzvppd TorsionDriv...
LRC-WPBEH/dzvp          OpenFF model LRC-WPBEH/dzvp TorsionDrive explo...
LRC-WPBEH/aug-cc-pvdz   OpenFF model LRC-WPBEH/aug-cc-pvdz TorsionDriv...
LRC-WPBEH/aug-cc-pvtz   OpenFF model LRC-WPBEH/aug-cc-pvtz TorsionDriv...
LRC-WPBEH/def2-tzvppd   OpenFF model LRC-WPBEH/def2-tzvppd TorsionDriv...
MN15-d3bJ/dzvp          OpenFF model MN15-d3bJ/dzvp TorsionDrive explo...
MN15-d3bJ/aug-cc-pvdz   OpenFF model MN15-d3bJ/aug-cc-pvdz TorsionDriv...
MN15-d3bJ/aug-cc-pvtz   OpenFF model MN15-d3bJ/aug-cc-pvtz TorsionDriv...
MN15-d3bJ/def2-tzvppd   OpenFF model MN15-d3bJ/def2-tzvppd TorsionDriv...
TPSSH-d3bJ/dzvp         OpenFF model TPSSH-d3bJ/dzvp TorsionDrive expl...
TPSSH-d3bJ/aug-cc-pvdz  OpenFF model TPSSH-d3bJ/aug-cc-pvdz TorsionDri...
TPSSH-d3bJ/aug-cc-pvtz  OpenFF model TPSSH-d3bJ/aug-cc-pvtz TorsionDri...
TPSSH-d3bJ/def2-tzvppd  OpenFF model TPSSH-d3bJ/def2-tzvppd TorsionDri...

In [7]:
res = get_incomplete_torsiondrive_optimizations(ds, 'default', client, merged=False)
res

{'20548975': [OptimizationRecord(id='20549211', status='COMPLETE'),
  OptimizationRecord(id='20550249', status='COMPLETE'),
  OptimizationRecord(id='20550250', status='COMPLETE'),
  OptimizationRecord(id='20556649', status='COMPLETE'),
  OptimizationRecord(id='20556650', status='COMPLETE'),
  OptimizationRecord(id='20556651', status='COMPLETE'),
  OptimizationRecord(id='20556652', status='COMPLETE'),
  OptimizationRecord(id='20657510', status='COMPLETE'),
  OptimizationRecord(id='20657511', status='COMPLETE'),
  OptimizationRecord(id='20657512', status='COMPLETE'),
  OptimizationRecord(id='20657513', status='COMPLETE'),
  OptimizationRecord(id='20658692', status='COMPLETE'),
  OptimizationRecord(id='20658693', status='COMPLETE'),
  OptimizationRecord(id='20658694', status='COMPLETE'),
  OptimizationRecord(id='20658695', status='COMPLETE'),
  OptimizationRecord(id='20659679', status='COMPLETE'),
  OptimizationRecord(id='20659680', status='COMPLETE'),
  OptimizationRecord(id='20659681', 

In [8]:
res.keys()

dict_keys(['20548975', '19712170', '20548997', '20548998', '19712173'])

In [9]:
optr = res['19712173'][-6]

In [10]:
optr.dict()

{'id': '20321335',
 'hash_index': '4c50828aa750287753274d21d0f14248989141a2',
 'procedure': 'optimization',
 'program': 'geometric',
 'version': 1,
 'protocols': {},
 'extras': {},
 'stdout': '23627523',
 'stderr': None,
 'error': None,
 'task_id': None,
 'manager_name': None,
 'status': <RecordStatusEnum.incomplete: 'INCOMPLETE'>,
 'modified_on': datetime.datetime(2020, 5, 14, 16, 52, 38, 501432),
 'created_on': datetime.datetime(2020, 5, 14, 2, 22, 4, 348635),
 'provenance': {'creator': 'geomeTRIC',
  'version': '0.9.7.2',
  'routine': 'geometric.run_json.geometric_run_json',
  'hostname': 'lt21',
  'wall_time': 1576.063833475113,
  'username': 'chodera',
  'qcengine_version': '0.14.0',
  'cpu': 'Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz'},
 'schema_version': 1,
 'initial_molecule': '13855947',
 'qc_spec': {'driver': <DriverEnum.gradient: 'gradient'>,
  'method': 'b3lyp-d3bj',
  'basis': 'dzvp',
  'keywords': '2',
  'program': 'psi4'},
 'keywords': {'coordsys': 'tric',
  'enforce': 0

In [11]:
ds.df

default
Cc1c[c:2]([c:1]2ccccc2c1OP(=O)([O-])[O-])[O:3][...  TorsionDriveRecord(id='19712153', status='COMP...
c1cc[c:1]2c(c1)ccc[c:2]2[O:3][P:4](=O)([O-])[O-]    TorsionDriveRecord(id='20548975', status='RUNN...
c1ccc(cc1)C[CH2:1][NH:2]/[C:3](=[NH+:4]/C(=[NH2...  TorsionDriveRecord(id='19712155', status='COMP...
C[CH2:1][NH:2]/[C:3](=[NH+:4]/C=[NH2+])/N           TorsionDriveRecord(id='20548976', status='COMP...
C[CH2:1][NH:2]/[C:3](=[NH+:4]/C)/N                  TorsionDriveRecord(id='20548977', status='COMP...
Cc1c([nH]cn1)CSC[CH2:1][NH:2]/[C:3](=N\C#N)/[NH...  TorsionDriveRecord(id='19712158', status='COMP...
C[CH2:1][NH:2]/[C:3](=[N:4]\C#N)/NC                 TorsionDriveRecord(id='20548978', status='COMP...
C[CH2:1][NH:2]/[C:3](=[N:4]\C)/NC                   TorsionDriveRecord(id='20548979', status='COMP...
c1cc(oc1)[CH2:1][NH:2][c:3]2cc(c(c[c:4]2C(=O)[O...  TorsionDriveRecord(id='19712161', status='COMP...
c1cc(oc1)[CH2:4][NH:3][c:2]2cc(c(c[c:1]2C(=O)[O...  TorsionDriveRecord(id='19712162', status='COMP...
c1cc[c:2]([c:1](c1)C(=O)[O-])[NH:3][CH2:4]c2ccco2   TorsionDriveRecord(id='20548992', status='COMP...
c1cc(oc1)CNc2c[c:1]([c:2](cc2C(=O)[O-])[S:3](=[...  TorsionDriveRecord(id='19712164', status='COMP...
c1c[c:1]([c:2](cc1C(=O)[O-])[S:3](=[O:4])(=O)N)Cl   TorsionDriveRecord(id='20548993', status='COMP...
c1cc[c:1]([c:2](c1)[S:3](=[O:4])(=O)N)Cl            TorsionDriveRecord(id='20548994', status='COMP...
CC[NH+](CC)C[CH2:4][O:3][C:2](=[O:1])c1ccc(cc1Cl)N  TorsionDriveRecord(id='19712167', status='COMP...
c1ccc(c(c1)[C:3](=[O:4])[O:2][CH2:1]C[NH3+])Cl      TorsionDriveRecord(id='20548995', status='COMP...
C[CH2:1][O:2][C:3](=[O:4])c1ccccc1Cl                TorsionDriveRecord(id='20548996', status='COMP...
c1c(cc([c:2]([c:1]1I)[O:3][c:4]2cc(c(c(c2)I)[O-...  TorsionDriveRecord(id='19712170', status='ERROR')
Cc1c[c:1]([c:2](c(c1)I)[O:3][c:4]2cc(c(c(c2)I)[...  TorsionDriveRecord(id='20548997', status='ERROR')
c1cc[c:1](cc1)[O:2][c:3]2c(ccc[c:4]2I)I             TorsionDriveRecord(id='20548998', status='ERROR')
CC(=O)Oc1ccccc1[C:1](=O)[NH:2][c:3]2ncc([s:4]2)...  TorsionDriveRecord(id='19712173', status='RUNN...
CC(=O)Oc1ccccc1[C:1](=O)[NH:2][c:3]2ncc[s:4]2       TorsionDriveRecord(id='20548999', status='COMP...

------------

In [12]:
res = get_incomplete_torsiondrive_optimizations(ds, 'default', client, merged=True)
res

[OptimizationRecord(id='19819414', status='ERROR'),
 OptimizationRecord(id='19820834', status='COMPLETE'),
 OptimizationRecord(id='19862575', status='COMPLETE'),
 OptimizationRecord(id='19862576', status='COMPLETE'),
 OptimizationRecord(id='19863476', status='COMPLETE'),
 OptimizationRecord(id='19863477', status='COMPLETE'),
 OptimizationRecord(id='19863478', status='COMPLETE'),
 OptimizationRecord(id='19863479', status='COMPLETE'),
 OptimizationRecord(id='19881891', status='COMPLETE'),
 OptimizationRecord(id='19881892', status='COMPLETE'),
 OptimizationRecord(id='19881893', status='COMPLETE'),
 OptimizationRecord(id='19881894', status='COMPLETE'),
 OptimizationRecord(id='19881895', status='COMPLETE'),
 OptimizationRecord(id='19881896', status='COMPLETE'),
 OptimizationRecord(id='20160968', status='COMPLETE'),
 OptimizationRecord(id='20160969', status='COMPLETE'),
 OptimizationRecord(id='20160970', status='COMPLETE'),
 OptimizationRecord(id='20160971', status='COMPLETE'),
 Optimization

In [13]:
optrs = [optr for optr in res if optr.status == 'INCOMPLETE']

In [14]:
out = map(lambda x: x.dict(), optrs)

In [18]:
next(out)

StopIteration: 

In [21]:
errors = (i.get_error().error_message for i in res if i.status == 'ERROR')

In [22]:
incompletes = tuple(i for i in res if i.status == 'INCOMPLETE')

In [23]:
incompletes[0].dict()

{'id': '19819414',
 'hash_index': '389f8f1518f247cee65989a5193d05f49df9900e',
 'procedure': 'optimization',
 'program': 'geometric',
 'version': 1,
 'protocols': {},
 'extras': {},
 'stdout': None,
 'stderr': None,
 'error': None,
 'task_id': None,
 'manager_name': None,
 'status': <RecordStatusEnum.incomplete: 'INCOMPLETE'>,
 'modified_on': datetime.datetime(2020, 6, 2, 0, 51, 57, 83768),
 'created_on': datetime.datetime(2020, 5, 5, 5, 30, 10, 933144),
 'provenance': None,
 'schema_version': 1,
 'initial_molecule': '13285322',
 'qc_spec': {'driver': <DriverEnum.gradient: 'gradient'>,
  'method': 'b3lyp-d3bj',
  'basis': 'dzvp',
  'keywords': '2',
  'program': 'psi4'},
 'keywords': {'coordsys': 'tric',
  'enforce': 0.1,
  'reset': True,
  'qccnv': True,
  'epsilon': 0,
  'constraints': {'set': [{'type': 'dihedral',
     'indices': [8, 7, 19, 6],
     'value': -90}]},
  'program': 'psi4'},
 'energies': None,
 'final_molecule': None,
 'trajectory': None}

### Discussion with Trevor

Need to mark the incompletes here as errored.

Also need to view input molecule.
There is no trajectory for this molecule.

Check to make sure indices are actually a torsion (input validation); indices are 0-based.

In [26]:
client.modify_tasks(operation='restart', base_result='19819414')

Data(n_updated=0)

### What molecules do we have failures or incompletes for?

In [12]:
ds = client.get_collection("TorsionDriveDataset", "OpenFF Fragmenter Validation 1.0")

In [13]:
ds.list_specifications()

Description
Name                                                                     
default                       Standard OpenFF torsiondrive specification.
B3LYP-d3bj/aug-cc-pvdz  OpenFF model B3LYP-d3bj/aug-cc-pvdz TorsionDri...
B3LYP-d3bj/aug-cc-pvtz  OpenFF model B3LYP-d3bj/aug-cc-pvtz TorsionDri...
B3LYP-d3bj/def2-tzvppd  OpenFF model B3LYP-d3bj/def2-tzvppd TorsionDri...
BLYP-d3bj/dzvp          OpenFF model BLYP-d3bj/dzvp TorsionDrive explo...
BLYP-d3bj/aug-cc-pvdz   OpenFF model BLYP-d3bj/aug-cc-pvdz TorsionDriv...
BLYP-d3bj/aug-cc-pvtz   OpenFF model BLYP-d3bj/aug-cc-pvtz TorsionDriv...
BLYP-d3bj/def2-tzvppd   OpenFF model BLYP-d3bj/def2-tzvppd TorsionDriv...
LRC-WPBEH/dzvp          OpenFF model LRC-WPBEH/dzvp TorsionDrive explo...
LRC-WPBEH/aug-cc-pvdz   OpenFF model LRC-WPBEH/aug-cc-pvdz TorsionDriv...
LRC-WPBEH/aug-cc-pvtz   OpenFF model LRC-WPBEH/aug-cc-pvtz TorsionDriv...
LRC-WPBEH/def2-tzvppd   OpenFF model LRC-WPBEH/def2-tzvppd TorsionDriv...
MN15-d3bJ/dzvp          OpenFF model MN15-d3bJ/dzvp TorsionDrive explo...
MN15-d3bJ/aug-cc-pvdz   OpenFF model MN15-d3bJ/aug-cc-pvdz TorsionDriv...
MN15-d3bJ/aug-cc-pvtz   OpenFF model MN15-d3bJ/aug-cc-pvtz TorsionDriv...
MN15-d3bJ/def2-tzvppd   OpenFF model MN15-d3bJ/def2-tzvppd TorsionDriv...
TPSSH-d3bJ/dzvp         OpenFF model TPSSH-d3bJ/dzvp TorsionDrive expl...
TPSSH-d3bJ/aug-cc-pvdz  OpenFF model TPSSH-d3bJ/aug-cc-pvdz TorsionDri...
TPSSH-d3bJ/aug-cc-pvtz  OpenFF model TPSSH-d3bJ/aug-cc-pvtz TorsionDri...
TPSSH-d3bJ/def2-tzvppd  OpenFF model TPSSH-d3bJ/def2-tzvppd TorsionDri...

In [14]:
res = get_incomplete_torsiondrive_optimizations(ds, 'default', client, merged=True)

In [15]:
for i in res:
    if i.status == 'ERROR':
        print(i.dict())

{'id': '20549279', 'hash_index': '537c8d5688663d69f1c7b33178ef8d3108fd7897', 'procedure': 'optimization', 'program': 'geometric', 'version': 1, 'protocols': {}, 'extras': {}, 'stdout': None, 'stderr': None, 'error': '23815495', 'task_id': None, 'manager_name': 'lilac_multithread-ls11-fb6c1425-5b8a-4e8f-84b9-23c967232875', 'status': <RecordStatusEnum.error: 'ERROR'>, 'modified_on': datetime.datetime(2020, 5, 15, 19, 58, 48, 977269), 'created_on': datetime.datetime(2020, 5, 15, 17, 46, 23, 438928), 'provenance': None, 'schema_version': 1, 'initial_molecule': '13285323', 'qc_spec': {'driver': <DriverEnum.gradient: 'gradient'>, 'method': 'b3lyp-d3bj', 'basis': 'dzvp', 'keywords': '2', 'program': 'psi4'}, 'keywords': {'coordsys': 'tric', 'enforce': 0.1, 'reset': True, 'qccnv': True, 'epsilon': 0, 'constraints': {'set': [{'type': 'dihedral', 'indices': [8, 7, 14, 6], 'value': -90}]}, 'program': 'psi4'}, 'energies': None, 'final_molecule': None, 'trajectory': None}
{'id': '20820900', 'hash_in

In [18]:
ds.df[ds.df.default.apply(lambda x: x.status == 'ERROR')].index.values

array(['Cc1c[c:1]([c:2](c(c1)I)[O:3][c:4]2cc(c(c(c2)I)[O-])I)I',
       'c1cc[c:1](cc1)[O:2][c:3]2c(ccc[c:4]2I)I'], dtype=object)

In [19]:
ds.df[ds.df.default.apply(lambda x: x.status == 'RUNNING')].index.values.tolist()

['c1cc[c:1]2c(c1)ccc[c:2]2[O:3][P:4](=O)([O-])[O-]',
 'c1c(cc([c:2]([c:1]1I)[O:3][c:4]2cc(c(c(c2)I)[O-])I)I)C[C@H](C(=O)[O-])[NH3+]',
 'CC(=O)Oc1ccccc1[C:1](=O)[NH:2][c:3]2ncc([s:4]2)[N+](=O)[O-]']

## Failure analysis 2020.05.28

In [62]:
[i for i in ds.df.default if i.status == 'ERROR']

[TorsionDriveRecord(id='20548994', status='ERROR'),
 TorsionDriveRecord(id='20548997', status='ERROR'),
 TorsionDriveRecord(id='20548998', status='ERROR')]

In [63]:
ids = set(i.id for i in ds.df.default)

In [64]:
res = client.query_procedures(ids)

In [65]:
from collections import defaultdict

In [66]:
angle_optimizations = defaultdict(set)
for tdr in ds.df.default:
    if tdr.status == 'COMPLETE':
        continue
        
    for val in tdr.optimization_history.values():
        angle_optimizations[tdr.id].update(set(val))

In [67]:
angle_optimizations

defaultdict(set,
            {'20548975': {'20549211',
              '20550249',
              '20550250',
              '20556649',
              '20556650',
              '20556651',
              '20556652',
              '20657510',
              '20657511',
              '20657512',
              '20657513',
              '20658692',
              '20658693',
              '20658694',
              '20658695',
              '20659679',
              '20659680',
              '20659681',
              '20659682',
              '20660333',
              '20660334',
              '20660335',
              '20660336',
              '20661477',
              '20661478',
              '20661479',
              '20661480',
              '20662676',
              '20662691',
              '20662692',
              '20662693',
              '20824606',
              '20824607',
              '20824608',
              '20824609',
              '20824610',
              '20824611',
         

In [68]:
angle_optimizations_i = set()

for i in angle_optimizations.values():
    angle_optimizations_i.update(set(i))
    
angle_optimizations_i

{'19819414',
 '19820834',
 '19862575',
 '19862576',
 '19863476',
 '19863477',
 '19863478',
 '19863479',
 '19881891',
 '19881892',
 '19881893',
 '19881894',
 '19881895',
 '19881896',
 '20160968',
 '20160969',
 '20160970',
 '20160971',
 '20160972',
 '20160973',
 '20176050',
 '20176051',
 '20176052',
 '20176053',
 '20285428',
 '20285429',
 '20285444',
 '20285445',
 '20295421',
 '20295422',
 '20295423',
 '20295424',
 '20321335',
 '20321336',
 '20321353',
 '20321354',
 '20321355',
 '20321356',
 '20549211',
 '20549256',
 '20549279',
 '20549280',
 '20550249',
 '20550250',
 '20555532',
 '20555533',
 '20556649',
 '20556650',
 '20556651',
 '20556652',
 '20557841',
 '20557842',
 '20562156',
 '20562157',
 '20562158',
 '20562159',
 '20563392',
 '20563393',
 '20563394',
 '20563395',
 '20570923',
 '20570924',
 '20570925',
 '20570926',
 '20570972',
 '20570973',
 '20573730',
 '20573731',
 '20573740',
 '20573741',
 '20576181',
 '20576223',
 '20576224',
 '20576225',
 '20576226',
 '20576227',
 '20627446',

In [69]:
len(angle_optimizations_i)

146

In [70]:
res_angle_opt = client.query_procedures(angle_optimizations_i)

In [71]:
res_angle_opt

[OptimizationRecord(id='19819414', status='INCOMPLETE'),
 OptimizationRecord(id='19820834', status='COMPLETE'),
 OptimizationRecord(id='19862575', status='COMPLETE'),
 OptimizationRecord(id='19862576', status='COMPLETE'),
 OptimizationRecord(id='19863476', status='COMPLETE'),
 OptimizationRecord(id='19863477', status='COMPLETE'),
 OptimizationRecord(id='19863478', status='COMPLETE'),
 OptimizationRecord(id='19863479', status='COMPLETE'),
 OptimizationRecord(id='19881891', status='COMPLETE'),
 OptimizationRecord(id='19881892', status='COMPLETE'),
 OptimizationRecord(id='19881893', status='COMPLETE'),
 OptimizationRecord(id='19881894', status='COMPLETE'),
 OptimizationRecord(id='19881895', status='COMPLETE'),
 OptimizationRecord(id='19881896', status='COMPLETE'),
 OptimizationRecord(id='20160968', status='COMPLETE'),
 OptimizationRecord(id='20160969', status='COMPLETE'),
 OptimizationRecord(id='20160970', status='COMPLETE'),
 OptimizationRecord(id='20160971', status='COMPLETE'),
 Optimiz

In [78]:
review = ((i, i.get_error().error_message) for i in res_angle_opt if i.status == 'ERROR')

In [82]:
print(next(review))

StopIteration: 

Perhaps a memory issue?

In [56]:
client.modify_tasks(operation='restart', base_result='20576181')

Data(n_updated=0)

In [36]:
for i in ds.df.default:
    if i.status == 'ERROR':
        print(i.optimization_history)

{'[-150]': ['20549256', '20563392', '20563393', '20576181'], '[-165]': ['20557841', '20573730'], '[-135]': ['20557842', '20573731'], '[-120]': ['20563394', '20576223'], '[180]': ['20563395', '20576226', '20576227'], '[-105]': ['20573740'], '[165]': ['20573741'], '[-90]': ['20576224'], '[150]': ['20576225']}
{'[-90]': ['20549279']}
{'[90]': ['20549280', '20562157', '20562158', '20627449'], '[75]': ['20555532', '20570924', '20570925'], '[105]': ['20555533', '20570926', '20570972'], '[60]': ['20562156', '20627447', '20627448'], '[120]': ['20562159', '20627450'], '[45]': ['20570923', '20820900'], '[135]': ['20570973', '20820901'], '[30]': ['20627446'], '[150]': ['20627459'], '[15]': ['20820899'], '[165]': ['20820902']}


In [37]:
failed = tuple(i.id for i in res_angle_opt if i.status == 'ERROR')

In [38]:
failed

('20549279', '20576181', '20820900', '20820902')

In [39]:
for i in ds.df.default:
    if i.status == 'ERROR':
        
        for key, value in i.optimization_history.items():
            for f in failed:
                if f in value:
                    print(key)

[-150]
[-90]
[45]
[165]


Restart again?

In [40]:
angle_optimizations.keys()

dict_keys(['20548975', '20548994', '19712170', '20548997', '20548998', '19712173'])

In [68]:
for tdrid in angle_optimizations.keys():
    client.modify_services(operation='restart', procedure_id=tdrid)

In [85]:
ds.status('default')

default
COMPLETE       16
ERROR           3
RUNNING         3

## Problem cases 2020.05.27


From 'OpenFF Fragmenter Validation 1.0', job index: c1ccc:1Cl job id: '20548994'. Error message:
```
ComputeError(error_type='CancelledError', error_message='Caught Executor Error:\nTraceback (most recent call last):\n  File "/home/chodera/miniconda/envs/qcfractal/lib/python3.6/site-packages/qcfractal/queue/executor_adapter.py", line 15, in _get_future\n    return future.result()\n  File "/home/chodera/miniconda/envs/qcfractal/lib/python3.6/concurrent/futures/_base.py", line 423, in result\n    raise CancelledError()\nconcurrent.futures._base.CancelledError\n')
)
```

In [27]:
erred = [i for i in ds.df.default.values if i.status == 'ERROR']
erred

[TorsionDriveRecord(id='20548994', status='ERROR'),
 TorsionDriveRecord(id='20548997', status='ERROR'),
 TorsionDriveRecord(id='20548998', status='ERROR')]

Start with the first one:

In [28]:
tdr = erred[0]

In [29]:
tdr.id

'20548994'

In [30]:
tdr.dict()

{'id': '20548994',
 'hash_index': 'b84cdd6e0d9d10f46169aebdc95dc63047f60939',
 'procedure': 'torsiondrive',
 'program': 'torsiondrive',
 'version': 1,
 'protocols': None,
 'extras': {},
 'stdout': None,
 'stderr': None,
 'error': None,
 'task_id': None,
 'manager_name': None,
 'status': <RecordStatusEnum.error: 'ERROR'>,
 'modified_on': datetime.datetime(2020, 5, 15, 17, 44, 23, 194510),
 'created_on': datetime.datetime(2020, 5, 15, 17, 44, 23, 194512),
 'provenance': {'creator': 'torsiondrive',
  'version': 'v0.9.8.1',
  'routine': 'torsiondrive.td_api'},
 'initial_molecule': ['13285318'],
 'keywords': {'dihedrals': [(5, 4, 9, 7)],
  'grid_spacing': [15],
  'dihedral_ranges': None,
  'energy_decrease_thresh': None,
  'energy_upper_limit': 0.1},
 'optimization_spec': {'program': 'geometric',
  'keywords': {'coordsys': 'tric',
   'enforce': 0.1,
   'reset': True,
   'qccnv': True,
   'epsilon': 0}},
 'qc_spec': {'driver': <DriverEnum.gradient: 'gradient'>,
  'method': 'b3lyp-d3bj',
  'b

In [31]:
client.modify_services(operation='restart', procedure_id=tdr.id)

Data(n_updated=1)

In [32]:
ds = client.get_collection("TorsionDriveDataset", "OpenFF Fragmenter Validation 1.0")

In [33]:
ds.list_specifications()

Description
Name                                                                     
default                       Standard OpenFF torsiondrive specification.
B3LYP-d3bj/aug-cc-pvdz  OpenFF model B3LYP-d3bj/aug-cc-pvdz TorsionDri...
B3LYP-d3bj/aug-cc-pvtz  OpenFF model B3LYP-d3bj/aug-cc-pvtz TorsionDri...
B3LYP-d3bj/def2-tzvppd  OpenFF model B3LYP-d3bj/def2-tzvppd TorsionDri...
BLYP-d3bj/dzvp          OpenFF model BLYP-d3bj/dzvp TorsionDrive explo...
BLYP-d3bj/aug-cc-pvdz   OpenFF model BLYP-d3bj/aug-cc-pvdz TorsionDriv...
BLYP-d3bj/aug-cc-pvtz   OpenFF model BLYP-d3bj/aug-cc-pvtz TorsionDriv...
BLYP-d3bj/def2-tzvppd   OpenFF model BLYP-d3bj/def2-tzvppd TorsionDriv...
LRC-WPBEH/dzvp          OpenFF model LRC-WPBEH/dzvp TorsionDrive explo...
LRC-WPBEH/aug-cc-pvdz   OpenFF model LRC-WPBEH/aug-cc-pvdz TorsionDriv...
LRC-WPBEH/aug-cc-pvtz   OpenFF model LRC-WPBEH/aug-cc-pvtz TorsionDriv...
LRC-WPBEH/def2-tzvppd   OpenFF model LRC-WPBEH/def2-tzvppd TorsionDriv...
MN15-d3bJ/dzvp          OpenFF model MN15-d3bJ/dzvp TorsionDrive explo...
MN15-d3bJ/aug-cc-pvdz   OpenFF model MN15-d3bJ/aug-cc-pvdz TorsionDriv...
MN15-d3bJ/aug-cc-pvtz   OpenFF model MN15-d3bJ/aug-cc-pvtz TorsionDriv...
MN15-d3bJ/def2-tzvppd   OpenFF model MN15-d3bJ/def2-tzvppd TorsionDriv...
TPSSH-d3bJ/dzvp         OpenFF model TPSSH-d3bJ/dzvp TorsionDrive expl...
TPSSH-d3bJ/aug-cc-pvdz  OpenFF model TPSSH-d3bJ/aug-cc-pvdz TorsionDri...
TPSSH-d3bJ/aug-cc-pvtz  OpenFF model TPSSH-d3bJ/aug-cc-pvtz TorsionDri...
TPSSH-d3bJ/def2-tzvppd  OpenFF model TPSSH-d3bJ/def2-tzvppd TorsionDri...

'default' always means 'b3lyp-d3bj' for us.

In [35]:
ds.status("default")

default
COMPLETE       16
RUNNING         4
ERROR           2

In [36]:
ds.df

default
Cc1c[c:2]([c:1]2ccccc2c1OP(=O)([O-])[O-])[O:3][...  TorsionDriveRecord(id='19712153', status='COMP...
c1cc[c:1]2c(c1)ccc[c:2]2[O:3][P:4](=O)([O-])[O-]    TorsionDriveRecord(id='20548975', status='RUNN...
c1ccc(cc1)C[CH2:1][NH:2]/[C:3](=[NH+:4]/C(=[NH2...  TorsionDriveRecord(id='19712155', status='COMP...
C[CH2:1][NH:2]/[C:3](=[NH+:4]/C=[NH2+])/N           TorsionDriveRecord(id='20548976', status='COMP...
C[CH2:1][NH:2]/[C:3](=[NH+:4]/C)/N                  TorsionDriveRecord(id='20548977', status='COMP...
Cc1c([nH]cn1)CSC[CH2:1][NH:2]/[C:3](=N\C#N)/[NH...  TorsionDriveRecord(id='19712158', status='COMP...
C[CH2:1][NH:2]/[C:3](=[N:4]\C#N)/NC                 TorsionDriveRecord(id='20548978', status='COMP...
C[CH2:1][NH:2]/[C:3](=[N:4]\C)/NC                   TorsionDriveRecord(id='20548979', status='COMP...
c1cc(oc1)[CH2:1][NH:2][c:3]2cc(c(c[c:4]2C(=O)[O...  TorsionDriveRecord(id='19712161', status='COMP...
c1cc(oc1)[CH2:4][NH:3][c:2]2cc(c(c[c:1]2C(=O)[O...  TorsionDriveRecord(id='19712162', status='COMP...
c1cc[c:2]([c:1](c1)C(=O)[O-])[NH:3][CH2:4]c2ccco2   TorsionDriveRecord(id='20548992', status='COMP...
c1cc(oc1)CNc2c[c:1]([c:2](cc2C(=O)[O-])[S:3](=[...  TorsionDriveRecord(id='19712164', status='COMP...
c1c[c:1]([c:2](cc1C(=O)[O-])[S:3](=[O:4])(=O)N)Cl   TorsionDriveRecord(id='20548993', status='COMP...
c1cc[c:1]([c:2](c1)[S:3](=[O:4])(=O)N)Cl            TorsionDriveRecord(id='20548994', status='RUNN...
CC[NH+](CC)C[CH2:4][O:3][C:2](=[O:1])c1ccc(cc1Cl)N  TorsionDriveRecord(id='19712167', status='COMP...
c1ccc(c(c1)[C:3](=[O:4])[O:2][CH2:1]C[NH3+])Cl      TorsionDriveRecord(id='20548995', status='COMP...
C[CH2:1][O:2][C:3](=[O:4])c1ccccc1Cl                TorsionDriveRecord(id='20548996', status='COMP...
c1c(cc([c:2]([c:1]1I)[O:3][c:4]2cc(c(c(c2)I)[O-...  TorsionDriveRecord(id='19712170', status='RUNN...
Cc1c[c:1]([c:2](c(c1)I)[O:3][c:4]2cc(c(c(c2)I)[...  TorsionDriveRecord(id='20548997', status='ERROR')
c1cc[c:1](cc1)[O:2][c:3]2c(ccc[c:4]2I)I             TorsionDriveRecord(id='20548998', status='ERROR')
CC(=O)Oc1ccccc1[C:1](=O)[NH:2][c:3]2ncc([s:4]2)...  TorsionDriveRecord(id='19712173', status='RUNN...
CC(=O)Oc1ccccc1[C:1](=O)[NH:2][c:3]2ncc[s:4]2       TorsionDriveRecord(id='20548999', status='COMP...